armsrc/mifaresim.c

```diff
@@ -0,0 +1,631 @@
+	uint8_t receivedCmd_par[MAX_MIFARE_PARITY_SIZE];
+	uint16_t receivedCmd_len;
+	uint8_t response[MAX_MIFARE_FRAME_SIZE];
+	uint8_t response_par[MAX_MIFARE_PARITY_SIZE];
+	
+	uint8_t rAUTH_NT[] = {0x01, 0x02, 0x03, 0x04};
+	uint8_t rAUTH_AT[] = {0x00, 0x00, 0x00, 0x00};
+		
+	//Here, we collect UID,sector,keytype,NT,AR,NR,NT2,AR2,NR2
+	// This will be used in the reader-only attack.
+
+	//allow collecting up to 7 sets of nonces to allow recovery of up to 7 keys
+	#define ATTACK_KEY_COUNT 7 // keep same as define in cmdhfmf.c -> readerAttack() (Cannot be more than 7)
+	nonces_t ar_nr_resp[ATTACK_KEY_COUNT*2]; //*2 for 2 separate attack types (nml, moebius) 36 * 7 * 2 bytes = 504 bytes
+	memset(ar_nr_resp, 0x00, sizeof(ar_nr_resp));
+
+	uint8_t ar_nr_collected[ATTACK_KEY_COUNT*2]; //*2 for 2nd attack type (moebius)
+	memset(ar_nr_collected, 0x00, sizeof(ar_nr_collected));
+	uint8_t	nonce1_count = 0;
+	uint8_t	nonce2_count = 0;
+	uint8_t	moebius_n_count = 0;
+	bool gettingMoebius = false;
+	uint8_t	mM = 0; //moebius_modifier for collection storage
+
+	// Authenticate response - nonce
+	uint32_t nonce;
+	if (flags & FLAG_RANDOM_NONCE) {
+		nonce = prand();
+	} else {
+		nonce = bytes_to_num(rAUTH_NT, 4);
+	}
+
+	// free eventually allocated BigBuf memory but keep Emulator Memory
+	BigBuf_free_keep_EM();
+
+	MifareSimInit(flags, datain, &responses, &cuid, &uid_len);
+	
+	// We need to listen to the high-frequency, peak-detected path.
+	iso14443a_setup(FPGA_HF_ISO14443A_TAGSIM_LISTEN);
+
+	// clear trace
+	clear_trace();
+	set_tracing(true);
+	ResetSspClk();
+	
+	bool finished = false;
+	bool button_pushed = BUTTON_PRESS();
+	int cardSTATE = MFEMUL_NOFIELD;
+
+	while (!button_pushed && !finished && !usb_poll_validate_length()) {
+		WDT_HIT();
+
+		// find reader field
+		if (cardSTATE == MFEMUL_NOFIELD) {
+			int vHf = (MAX_ADC_HF_VOLTAGE * AvgAdc(ADC_CHAN_HF)) >> 10;
+			if (vHf > MF_MINFIELDV) {
+				LED_A_ON();
+				cardSTATE_TO_IDLE();
+			}
+			button_pushed = BUTTON_PRESS();
+			continue;
+		}
+
+		//Now, get data
+		int res = EmGetCmd(receivedCmd, &receivedCmd_len, receivedCmd_par);
+		if (res == 2) { //Field is off!
+			LEDsoff();
+			cardSTATE = MFEMUL_NOFIELD;
+			continue;
+		} else if (res == 1) { // button pressed
+			button_pushed = true;
+			break;
+		}
+
+		// WUPA in HALTED state or REQA or WUPA in any other state
+		if (receivedCmd_len == 1 && ((receivedCmd[0] == ISO14443A_CMD_REQA && cardSTATE != MFEMUL_HALTED) || receivedCmd[0] == ISO14443A_CMD_WUPA)) {
+			EmSendPrecompiledCmd(&responses[ATQA], (receivedCmd[0] == ISO14443A_CMD_WUPA));
+
+			// init crypto block
+			crypto1_destroy(pcs);
+			cardAUTHKEY = 0xff;
+			if (flags & FLAG_RANDOM_NONCE) {
+				nonce = prand();
+			}
+			LED_B_OFF();
+			LED_C_OFF();
+			cardSTATE = MFEMUL_SELECT1;
+			continue;
+		}
+		
+		switch (cardSTATE) {
+			case MFEMUL_NOFIELD:
+			case MFEMUL_HALTED:
+			case MFEMUL_IDLE:{
+				EmLogTraceReader();
+				break;
+			}
+			case MFEMUL_SELECT1:{
```

Comment:
looks there is an old bug with SELECT (// select cl2 card - 0x95 0x70 xxxxxxxxxxxx) without anticollision

Reply:
I don't see what you mean?

Reply:
there are 2 sequences to select:
1 - wupa - anticolision-select uid
2 - wupa - select uid

1st scenario works
2nd - none

Reply:
Both cases are handled. Can you provide a trace?

Reply:
No, sorry. Maybe I see reader's bug

Reply:
Maybe not. How do you know that the reader does select_uid without having a trace?

Reply:
command not works from reader. i dont look to trace. tomorrow ill check (im not near proxmark).

Reply:
all is ok
````
     Start |        End | Src | Data (! denotes parity error)                                   | CRC | Annotation         |     
------------|------------|-----|-----------------------------------------------------------------|-----|--------------------|     
          0 |       1056 | Rdr | 26                                                              |     | REQA
       2292 |       4660 | Tag | 04  00                                                          |     |
   38610936 |   38621400 | Rdr | 93  70  e6  84  87  f3  16  5e  35                              |  ok | SELECT_UID
   38622892 |   38626412 | Tag | 08  b6  dd                                                      |     |
````